cms/djangoapps/contentstore/views/tests/test_item.py

```diff
@@ -680,6 +681,152 @@ def verify_name(source_usage_key, parent_usage_key, expected_name, display_name=
+            'dest_source_locator': unicode(dest_usage_key)
+        }
+        if source_index is not None:
+            data['source_index'] = source_index
+        resp = self.client.ajax_post(reverse('contentstore.views.xblock_handler'), json.dumps(data))
+        return resp
+
+    def assert_move_item(self, source_usage_key, dest_usage_key, source_index=None):
+        """
+        Assert move component.
+
+        Arguments:
+            source_usage_key (BlockUsageLocator): Locator of source item.
+            dest_usage_key (BlockUsageLocator): Locator of destination item.
+            source_index (int): If provided, insert source item at the provided index location in dest_usage_key item.
+        """
+        parent_loc = self.store.get_parent_location(source_usage_key)
+        actual_souce_index = _get_source_index(source_usage_key)
+        response = self._move_component(source_usage_key, dest_usage_key, source_index)
+        self.assertEqual(response.status_code, 200)
+        response = json.loads(response.content)
+        self.assertEqual(response['move_source_locator'], unicode(source_usage_key))
+        self.assertEqual(response['dest_source_locator'], unicode(dest_usage_key))
+        source_index = source_index if source_index is not None else actual_souce_index
+        self.assertEqual(response['source_index'], source_index)
+        new_parent_loc = self.store.get_parent_location(source_usage_key)
+        self.assertEqual(new_parent_loc, dest_usage_key)
+        self.assertNotEqual(parent_loc, new_parent_loc)
+
+    def test_move_component(self):
+        """
+        Test move component with different xblock types.
+        """
+        for source_usage_key, dest_usage_key in [
+                (self.html_usage_key, self.vert2_usage_key),
+                (self.vert_usage_key, self.seq2_usage_key),
+                (self.seq_usage_key, self.chapter2_usage_key)
+        ]:
+            self.assert_move_item(source_usage_key, dest_usage_key)
+
+    def test_move_source_index(self):
+        """
+        Test moving an item to a particular index.
+        """
+        parent = self.get_item_from_modulestore(self.vert_usage_key)
+        children = parent.get_children()
+        self.assertEqual(len(children), 2)
+
+        # Create a component within vert2.
+        resp = self.create_xblock(parent_usage_key=self.vert2_usage_key, display_name='html2', category='html')
+        html2_usage_key = self.response_usage_key(resp)
+
+        # Move html2_usage_key inside vert_usage_key at second position.
+        self.assert_move_item(html2_usage_key, self.vert_usage_key, 1)
+        parent = self.get_item_from_modulestore(self.vert_usage_key)
+        children = parent.get_children()
+        self.assertEqual(len(children), 3)
+        self.assertEqual(children[1].location, html2_usage_key)
+
+    def test_move_undo(self):
+        """
+        Test move a component and move it back (undo).
```

Comment:
This isn't really an undo as it doesn't preserve the original source index (or does it somehow)?